### PR TITLE
Cross attention

### PR DESCRIPTION
**Pull request by stepanyanhayk**
Implements Cross-Attention Fusion (medium capacity) — Issue #[31
](https://github.com/McDermottHealthAI/MedRAP/issues/31)                                                                                             
  New module — CrossAttentionFusion (fusion.py)
  Transformer decoder-style fusion where the patient sequence cross-attends to retrieved     
  document tokens. Stacks N layers of cross-attention → residual+LayerNorm → GELU FFN →      
  residual+LayerNorm. Supports padding masks from doc_attention_mask. Fully doctested.       
                                                                                             
  New configs                                                                                
  - conf/fusion/cross_attention_medium.yaml — 2 layers, d_model=256, 8 heads, ff_dim=512
  - conf/retrieval_encoder/token_feature.yaml — added vocab_size and embedding_dim as        
  declared fields (Qwen3 defaults)
                                                                                             
  Training improvements
  - BinaryClassificationLoss now accepts pos_weight for class-imbalanced datasets            
  - MedRAPSupervisedLightningModule now supports cosine annealing LR scheduler with linear   
  warmup via warmup_steps                                                                 
  - gradient_clip_val added as a declared field to both trainer configs                      
                  
  Cluster script — scripts/train_cross_attention_medium_slurm.sh                             
  End-to-end SLURM script for the medium config on MIMIC: token_embedding_128 encoder,
  token_feature retrieval encoder (K=8, max_seq_len=256), masked_mean pooling, binary BCE    
  with pos_weight=9.0, LR=1e-4 with 200-step warmup, gradient clipping, auto-upload to GCS on
   completion.                                                                               
   

https://wandb.ai/haykstepanyan02-columbia-university/medrap/runs/ams2j8jz?nw=nwuserhaykstepanyan02

### CODE REVIEW

File: scripts/train_cross_attention_medium_slurm.sh
Comment (mmcdermott) [inline scripts/train_cross_attention_medium_slurm.sh:1]: In general, though don't worry about this for now, these sorts of script files probably should be in a different repo, as they are cluster / dataset specific experimental scripts, rather than general model code.

File: src/medrap/conf/retrieval_encoder/token_feature.yaml [hunk 1]
```diff
@@ -1 +1,3 @@
 _target_: medrap.retrieval_encoder.TokenFeatureRetrievalEncoder
+vocab_size: 151936
```
Comment (mmcdermott) [inline src/medrap/conf/retrieval_encoder/token_feature.yaml]: Really, we don't want this hard coded, but instead should somehow be connected to the value (or key?) encoder we use for the retrieval process, depending on which this is for.

File: src/medrap/conf/fusion/cross_attention_medium.yaml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+_target_: medrap.fusion.CrossAttentionFusion
+d_model: 256
+num_heads: 8
+ff_dim: 512
+num_layers: 2
+d_in_patient: ???   # must match the patient encoder output dimension (D_ehr)
```
Comment (mmcdermott) [inline src/medrap/conf/fusion/cross_attention_medium.yaml]: If these must match, need they be separate parameters?

File: src/medrap/lightning_module.py [hunk 3]
```diff
@@ -370,4 +372,20 @@ def configure_optimizers(self) -> Optimizer:
...
         """
-        return self.optimizer_factory(self._grouped_parameters())
+        optimizer = self.optimizer_factory(self._grouped_parameters())
+        if self.warmup_steps == 0:
+            return optimizer
+        total_steps = self.trainer.estimated_stepping_batches
+        warmup = torch.optim.lr_scheduler.LinearLR(
+            optimizer, start_factor=1e-8, end_factor=1.0, total_iters=self.warmup_steps
+        )
+        cosine = torch.optim.lr_scheduler.CosineAnnealingLR(
```
Comment (mmcdermott) [inline src/medrap/lightning_module.py]: I think there are better options in HF to use here. It's a bit indirect, but check out what's done in MEDS-EIC-AR: https://github.com/mmcdermott/MEDS_EIC_AR/blob/main/src/MEDS_EIC_AR/training/module.py#L428 and https://github.com/mmcdermott/MEDS_EIC_AR/blob/main/src/MEDS_EIC_AR/configs/lightning_module/default.yaml and https://github.com/mmcdermott/MEDS_EIC_AR/blob/main/src/MEDS_EIC_AR/configs/lightning_module/LR_scheduler/get_cosine_schedule_with_warmup.yaml

File: src/medrap/task.py [hunk 4]
```diff
@@ -277,12 +277,21 @@ def metrics(self, predictions: ModelOutput, targets: TaskTargets) -> Mapping[str
 class BinaryClassificationLoss(SupervisedLoss):
     """Binary BCE-with-logits loss for scalar binary predictions.
 
+    Args:
+        pos_weight: Weight for positive class, useful for imbalanced datasets.
```
Comment (mmcdermott) [inline src/medrap/task.py]: Don't include this. This is not actually typically a good idea.

**Review by mmcdermott — COMMENTED** [review]
looks mostly good, but review this to make sure the cross attention architecture setup is good: https://chatgpt.com/share/69e63511-d5b0-83ea-96c5-956c26cc7428

**Review by zaczw — APPROVED** [review]
